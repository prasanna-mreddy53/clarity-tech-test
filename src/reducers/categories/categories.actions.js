import { categoryActionTypes } from "./categories.types";

export const fetchCategorieRequest = () => ({
  type: categoryActionTypes.FETCH_CATEGORIES_REQUEST,
});

export const fetchCategorieSuccess = (json) => ({
  type: categoryActionTypes.FETCH_CATEGORIES_SUCCESS,
  payload: json,
});

export const fetchCategorieFailure = (error) => ({
  type: categoryActionTypes.FETCH_CATEGORIES_FAILURE,
  payload: error,
});

export const fetchCategories = () => {
  return async (dispatch) => {
    dispatch(fetchCategorieRequest);
    try {
      let response = await fetch(
        `https://www.thecocktaildb.com/api/json/v1/1/list.php?c=list`
      );
      let json = await response.json();
      dispatch(fetchCategorieSuccess(json));
    } catch (error) {
      dispatch(fetchCategorieFailure(error));
    }
  };
};
