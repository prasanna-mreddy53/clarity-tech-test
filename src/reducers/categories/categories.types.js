export const categoryActionTypes = {
  FETCH_CATEGORIES_REQUEST: "FETCH_CATEGORIES_REQUEST",
  FETCH_CATEGORIES_SUCCESS: "FETCH_CATEGORIES_SUCCESS",
  FETCH_CATEGORIES_FAILURE: "FETCH_CATEGORIES_FAILURE",
};
