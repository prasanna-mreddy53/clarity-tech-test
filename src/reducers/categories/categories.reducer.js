import { categoryActionTypes } from "./categories.types";

const INITIAL_STATE = {
  loading: false,
  categories: [],
  error: "",
};

const categoriesReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case categoryActionTypes.FETCH_CATEGORIES_REQUEST:
      return {
        ...state,
        loading: true,
      };
    case categoryActionTypes.FETCH_CATEGORIES_FAILURE:
      return {
        ...state,
        error: action.payload,
      };
    case categoryActionTypes.FETCH_CATEGORIES_SUCCESS:
      return {
        ...state,
        categories: action.payload,
      };
    default:
      return state;
  }
};

export default categoriesReducer;
