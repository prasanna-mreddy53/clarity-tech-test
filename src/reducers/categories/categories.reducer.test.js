import reducer from "./categories.reducer";
import { categoryActionTypes } from "./categories.types";

describe("Categories reducer", () => {
  it("Categories should return the initial state", () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it("should return the initial state with default values", () => {
    expect(reducer(undefined, {})).toEqual({
      loading: false,
      categories: [],
      error: "",
    });
  });

  it("should match FETCH_CATEGORIES_REQUEST action", () => {
    expect(
      reducer([], {
        type: categoryActionTypes.FETCH_CATEGORIES_REQUEST,
      })
    ).toEqual({
      loading: true,
    });
  });

  it("should match FETCH_CATEGORIES_SUCCESS action", () => {
    expect(
      reducer(
        { categories: [], loading: false, error: "" },
        {
          type: categoryActionTypes.FETCH_CATEGORIES_SUCCESS,
          categories: { drinks: [{ strCategory: "Ordinary Drink" }] },
        }
      )
    ).toEqual({
      categories: [],
      loading: false,
      error: "",
      categories: undefined,
    });
  });

  it("Categories FETCH_CATEGORIES_SUCCESS action", () => {
    expect(
      reducer(
        {
          categories: { drinks: [{ strCategory: "Ordinary Drink" }] },
        },
        categoryActionTypes.FETCH_CATEGORIES_SUCCESS
      )
    ).toMatchSnapshot();
  });

  it("should match FETCH_CATEGORIES_FAILURE action", () => {
    expect(
      reducer([{ categories: [], loading: false, error: "" }], {
        type: categoryActionTypes.FETCH_CATEGORIES_FAILURE,
        error: "could not load categories",
      })
    ).toMatchSnapshot();
  });
});
