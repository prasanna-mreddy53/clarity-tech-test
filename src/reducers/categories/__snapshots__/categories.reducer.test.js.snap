// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories reducer Categories FETCH_CATEGORIES_SUCCESS action 1`] = `
Object {
  "categories": Object {
    "drinks": Array [
      Object {
        "strCategory": "Ordinary Drink",
      },
    ],
  },
}
`;

exports[`Categories reducer Categories should return the initial state 1`] = `
Object {
  "categories": Array [],
  "error": "",
  "loading": false,
}
`;

exports[`Categories reducer should match FETCH_CATEGORIES_FAILURE action 1`] = `
Object {
  "0": Object {
    "categories": Array [],
    "error": "",
    "loading": false,
  },
  "error": undefined,
}
`;
